lib/sqlitex.ex

```diff
@@ -7,19 +7,47 @@ defmodule Sqlitex do
     :esqlite3.open(path)
   end
 
-  def query(db, query, opts \\ []) do
-    {:ok, statement} = :esqlite3.prepare(query, db)
-    types = :esqlite3.column_types(statement) |> Tuple.to_list
-    columns = :esqlite3.column_names(statement) |> Tuple.to_list
-    rows = :esqlite3.fetchall(statement)
-    into = Keyword.get(opts, :into, [])
-    Sqlitex.Row.from(types, columns, rows, into)
-  end
-
   def with_db(path, fun) do
     {:ok, db} = open(path)
     res = fun.(db)
     close(db)
     res
   end
+
+  def exec(db, sql) do
+    :esqlite3.exec(sql, db)
+  end
+
+  def query(db, sql, opts \\ []) do
+    {params, into} = query_options(opts)
+    {:ok, statement} = :esqlite3.prepare(sql, db)
+    :ok = :esqlite3.bind(statement, params)
+    types = :esqlite3.column_types(statement)
+    columns = :esqlite3.column_names(statement)
+    rows = :esqlite3.fetchall(statement)
+    return_rows_or_error(types, columns, rows, into)
+  end
+
+  defp query_options(opts) do
+    params = Keyword.get(opts, :bind, [])
+    into = Keyword.get(opts, :into, [])
+    {params, into}
+  end
+
+  defp return_rows_or_error({:error, kind} = error, columns, rows, into) do
+    case kind do
+      :no_columns -> return_rows_or_error({}, columns, rows, into)
+      _ -> error
+    end
+  end
+  defp return_rows_or_error(types, {:error, kind} = error, rows, into) do
+    case kind do
+      :no_columns -> return_rows_or_error(types, {}, rows, into)
+      _ -> error
+    end
+  end
```

Comment:
I'll look into writing a fix for the esqlite code.

I kind of like having a separate `query` and `exec`.  The first implies that you care about the output while the latter is more like a fire-and-forget -- it only returns `:ok` or an error.  As a wrapper around `sqlite3_exec`, the `exec` call can't take parametrized statements.  In [sqlite_ecto](https://github.com/jazzyb/sqlite_ecto) I use `exec` internally for imperative commands like transactions and triggers -- things for which bound variables wouldn't make much sense anyway.

Ecto uses `query` because it first builds statements dynamically and then passes these SQL strings back to the adapter to be executed.  The adapter is agnostic about the statement being passed in so using `query` for everything is a safe bet.